src/internal/list.js

```diff
@@ -0,0 +1,2 @@
+export const empty = ({isEmpty: true, size: 0, head: null, tail: null});
+export const cons = (head, tail) => ({isEmpty: false, size: tail.size + 1, head, tail});
```

Comment:
Surprisingly, defining the list type like this has zero impact on performance. What do you think @davidchambers and @safareli? I think it addresses both of your points of feedback. :)